File: backend/internal/services/embedding_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"nexia-backend/internal/models"

	"go.uber.org/zap"
	"gorm.io/gorm"
)

// EmbeddingService is the background worker that converts profiles into vector embeddings
// and persists them to the profile_embeddings table for RAG retrieval.
type EmbeddingService struct {
	Profiles  ProfileRepository
	Generator EmbeddingGenerator
	Repo      EmbeddingRepository
	Logger    *zap.Logger
}

// EmbeddingGenerator generates a vector embedding from text.
type EmbeddingGenerator interface {
	GenerateEmbedding(ctx context.Context, text string) ([]float32, error)
}

// EmbeddingRepository persists and removes profile vector embeddings.
type EmbeddingRepository interface {
	UpsertProfile(ctx context.Context, profile *models.Profile, embedding []float32) error
	DeleteProfile(ctx context.Context, profileID uint64) error
}

// NewEmbeddingService constructs an EmbeddingService.
func NewEmbeddingService(profiles ProfileRepository, generator EmbeddingGenerator, repo EmbeddingRepository, logger *zap.Logger) *EmbeddingService {
	return &EmbeddingService{
		Profiles:  profiles,
		Generator: generator,
		Repo:      repo,
		Logger:    logger.Named("embedding_service"),
	}
}

// EmbedProfile loads a profile, generates its embedding, and upserts it into the vector store.
// Returns ErrNotFound if the profile no longer exists (caller should skip retry).
func (s *EmbeddingService) EmbedProfile(ctx context.Context, profileID uint) error {
	profile, err := s.Profiles.LoadForEmbedding(ctx, profileID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return ErrNotFound
		}
		return fmt.Errorf("fetching profile failed: %w", err)
	}

	text := buildEmbeddingText(profile)

	embedding, err := s.Generator.GenerateEmbedding(ctx, text)
	if err != nil {
		s.Logger.Error("embedding generation failed", zap.Uint("profile_id", profileID), zap.Error(err))
		return fmt.Errorf("gemini embedding failed: %w", err)
	}

	if err := s.Repo.UpsertProfile(ctx, profile, embedding); err != nil {
		s.Logger.Error("embedding upsert failed", zap.Uint("profile_id", profileID), zap.Error(err))
		return fmt.Errorf("pgvector upsert failed: %w", err)
	}

	s.Logger.Info("profile embedded",
		zap.Uint("profile_id", profileID),
		zap.Uint64("user_id", profile.UserID),
	)
	return nil
}

// DeleteEmbedding removes a profile's vector embedding from the store.
func (s *EmbeddingService) DeleteEmbedding(ctx context.Context, profileID uint64) error {
	return s.Repo.DeleteProfile(ctx, profileID)
}

// buildEmbeddingText flattens a fully-preloaded profile into a single string for embedding.
func buildEmbeddingText(profile *models.Profile) string {
	var sb strings.Builder
	fmt.Fprintf(&sb, "Profile of %s\n", profile.FullName)
	fmt.Fprintf(&sb, "Bio: %s\n", profile.Bio)
	fmt.Fprintf(&sb, "Profession: %s\n", profile.Profession)
	fmt.Fprintf(&sb, "Relationship Type: %s\n", profile.RelationshipType)
	if profile.ZodiacSign != nil {
		fmt.Fprintf(&sb, "Zodiac Sign: %s\n", *profile.ZodiacSign)
	}
	if profile.Birthday != nil {
		fmt.Fprintf(&sb, "Birthday: %s\n", time.Time(*profile.Birthday).Format("January 02, 2006"))
	}
	fmt.Fprintf(&sb, "Long Term Goals: %s\n", profile.LongTermGoals)
	fmt.Fprintf(&sb, "Music Preference: %s\n", profile.MusicPreference)
	fmt.Fprintf(&sb, "Favorite Movie: %s\n", profile.FavoriteMovie)
	fmt.Fprintf(&sb, "Favorite Book: %s\n", profile.FavoriteBook)
	fmt.Fprintf(&sb, "Favorite Memory: %s\n", profile.FavoriteMemory)

	if len(profile.Tags) > 0 {
		sb.WriteString("Interests/Tags: ")
		for i, t := range profile.Tags {
			sb.WriteString(t.Tag)
			if i < len(profile.Tags)-1 {
				sb.WriteString(", ")
			}
		}
		sb.WriteString("\n")
	}

	if len(profile.PoliticalViews) > 0 {
		sb.WriteString("Political Views: ")
		for i, v := range profile.PoliticalViews {
			sb.WriteString(v.View)
			if i < len(profile.PoliticalViews)-1 {
				sb.WriteString(", ")
			}
		}
		sb.WriteString("\n")
	}

	if len(profile.FoodRestrictions) > 0 {
		sb.WriteString("Food Restrictions: ")
		for i, f := range profile.FoodRestrictions {
			sb.WriteString(f.Restriction)
			if i < len(profile.FoodRestrictions)-1 {
				sb.WriteString(", ")
			}
		}
		sb.WriteString("\n")
	}

	if len(profile.MovieGenres) > 0 {
		sb.WriteString("Favorite Movie Genres: ")
		for i, g := range profile.MovieGenres {
			sb.WriteString(g.Genre)
			if i < len(profile.MovieGenres)-1 {
				sb.WriteString(", ")
			}
		}
		sb.WriteString("\n")
	}

	if len(profile.BookGenres) > 0 {
		sb.WriteString("Favorite Book Genres: ")
		for i, g := range profile.BookGenres {
			sb.WriteString(g.Genre)
			if i < len(profile.BookGenres)-1 {
				sb.WriteString(", ")
			}
		}
		sb.WriteString("\n")
	}

	if len(profile.HangoutPlaces) > 0 {
		sb.WriteString("Favorite Hangout Places: ")
		for i, p := range profile.HangoutPlaces {
			sb.WriteString(p.Place)
			if i < len(profile.HangoutPlaces)-1 {
				sb.WriteString(", ")
			}
		}
		sb.WriteString("\n")
	}

	if len(profile.TopSongs) > 0 {
		sb.WriteString("Top Songs: ")
		for i, song := range profile.TopSongs {
			fmt.Fprintf(&sb, "%s by %s", song.Name, song.Artist)
			if i < len(profile.TopSongs)-1 {
				sb.WriteString(", ")
			}
		}
		sb.WriteString("\n")
	}

	if profile.AssociatedSong != nil {
		fmt.Fprintf(&sb, "Associated Song: %s by %s\n", profile.AssociatedSong.Name, profile.AssociatedSong.Artist)
	}

	if len(profile.Quotes) > 0 {
		sb.WriteString("Quotes: ")
		for i, q := range profile.Quotes {
			fmt.Fprintf(&sb, "%q", q.Quote)
			if i < len(profile.Quotes)-1 {
				sb.WriteString(", ")
			}
		}
		sb.WriteString("\n")
	}

	return sb.String()
}

File: apps/api/src/services/embedding-service.ts
import type { ProfileOutput } from "@nexia/shared";
import type { Logger } from "../logging/logger";
import { errNotFound, errAIUnavailable } from "./errors";

export interface EmbeddingGenerator {
  generateEmbedding(text: string): Promise<number[]>;
}

export interface EmbeddingStorage {
  upsertProfile(
    profileId: number,
    userId: number,
    embedding: number[],
    payload: Record<string, unknown>
  ): Promise<void>;
  deleteProfile(profileId: number): Promise<void>;
}

export interface ProfileLoader {
  loadForEmbedding(profileId: number): Promise<ProfileOutput | null>;
}

export class EmbeddingService {
  constructor(
    private profiles: ProfileLoader,
    private generator: EmbeddingGenerator,
    private repo: EmbeddingStorage,
    private logger: Logger
  ) {}

  async embedProfile(profileId: number): Promise<void> {
    const profile = await this.profiles.loadForEmbedding(profileId);
    if (!profile) {
      throw errNotFound();
    }

    const text = buildEmbeddingText(profile);

    let embedding: number[];
    try {
      embedding = await this.generator.generateEmbedding(text);
    } catch (err) {
      this.logger.error({ profileId, err: String(err) }, "embedding generation failed");
      throw errAIUnavailable(
        `gemini embedding failed: ${err instanceof Error ? err.message : String(err)}`
      );
    }

    try {
      await this.repo.upsertProfile(profile.id, profile.user_id, embedding, profile);
    } catch (err) {
      this.logger.error({ profileId, err: String(err) }, "embedding upsert failed");
      throw errAIUnavailable(
        `pgvector upsert failed: ${err instanceof Error ? err.message : String(err)}`
      );
    }

    this.logger.info({ profileId, userId: profile.user_id }, "profile embedded");
  }

  async deleteEmbedding(profileId: number): Promise<void> {
    await this.repo.deleteProfile(profileId);
  }
}

const MONTHS = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
] as const;

/** Formats "YYYY-MM-DD" as "Month DD, YYYY" to match the Go embedding text. */
function formatDate(dateStr: string): string {
  const d = new Date(`${dateStr}T00:00:00Z`);
  const month = MONTHS[d.getUTCMonth()];
  const day = String(d.getUTCDate()).padStart(2, "0");
  return `${month} ${day}, ${d.getUTCFullYear()}`;
}

/** Flattens a profile into the labelled text block fed to the embedding model. */
export function buildEmbeddingText(p: ProfileOutput): string {
  const lines: string[] = [];
  lines.push(`Profile of ${p.full_name}`);
  lines.push(`Bio: ${p.bio}`);
  lines.push(`Profession: ${p.profession}`);
  lines.push(`Relationship Type: ${p.relationship_type}`);
  if (p.zodiac_sign) lines.push(`Zodiac Sign: ${p.zodiac_sign}`);
  if (p.birthday) lines.push(`Birthday: ${formatDate(p.birthday)}`);
  lines.push(`Long Term Goals: ${p.long_term_goals}`);
  lines.push(`Music Preference: ${p.music_preference}`);
  lines.push(`Favorite Movie: ${p.favorite_movie}`);
  lines.push(`Favorite Book: ${p.favorite_book}`);
  lines.push(`Favorite Memory: ${p.favorite_memory}`);
  lines.push(`Notes: ${p.notes}`);

  if (p.tags.length) lines.push(`Interests/Tags: ${p.tags.map((t) => t.tag).join(", ")}`);
  if (p.political_views.length) {
    lines.push(`Political Views: ${p.political_views.map((v) => v.view).join(", ")}`);
  }
  if (p.food_restrictions.length) {
    lines.push(`Food Restrictions: ${p.food_restrictions.map((r) => r.restriction).join(", ")}`);
  }
  if (p.movie_genres.length) {
    lines.push(`Favorite Movie Genres: ${p.movie_genres.map((g) => g.genre).join(", ")}`);
  }
  if (p.book_genres.length) {
    lines.push(`Favorite Book Genres: ${p.book_genres.map((g) => g.genre).join(", ")}`);
  }
  if (p.hangout_places.length) {
    lines.push(`Favorite Hangout Places: ${p.hangout_places.map((h) => h.place).join(", ")}`);
  }
  if (p.top_songs.length) {
    lines.push(`Top Songs: ${p.top_songs.map((s) => `${s.name} by ${s.artist}`).join(", ")}`);
  }
  if (p.associated_song) {
    lines.push(`Associated Song: ${p.associated_song.name} by ${p.associated_song.artist}`);
  }
  if (p.quotes.length) {
    lines.push(`Quotes: ${p.quotes.map((q) => `"${q.quote}"`).join(", ")}`);
  }

  return lines.join("\n");
}
